### PR TITLE
feat: Webhook for Role/RoleBinding

### PR DESCRIPTION
**Pull request by Arvindthiru**
### Description of your changes

Fixes #

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### CODE REVIEW

File: pkg/webhook/fleetresourcehandler/fleetresourcehandler_webhook.go [hunk 1]
```diff
@@ -41,24 +44,32 @@ type fleetResourceValidator struct {
...
 	var response admission.Response
 	if req.Operation == admissionv1.Create || req.Operation == admissionv1.Update || req.Operation == admissionv1.Delete {
 		switch req.Kind {
 		case createCRDGVK():
-			klog.V(2).InfoS("handling CRD resource", "GVK", createCRDGVK())
+			klog.V(2).InfoS("handling CRD resource", "GVK", createCRDGVK(), "Name/Namespace", fmt.Sprintf("%s/%s", req.Name, req.Namespace))
 			response = v.handleCRD(req)
 		case createMemberClusterGVK():
-			klog.V(2).InfoS("handling Member cluster resource", "GVK", createMemberClusterGVK())
+			klog.V(2).InfoS("handling Member cluster resource", "GVK", createMemberClusterGVK(), "Name/Namespace", fmt.Sprintf("%s/%s", req.Name, req.Namespace))
```
Comment (michaelawyu) [inline pkg/webhook/fleetresourcehandler/fleetresourcehandler_webhook.go:56]: Hi Arvind! A nit: we can just use `namespacedName` here if you'd like

File: pkg/webhook/fleetresourcehandler/fleetresourcehandler_webhook.go [hunk 2]
```diff
@@ -67,25 +78,47 @@ func (v *fleetResourceValidator) handleCRD(req admission.Request) admission.Resp
 
 	// This regex works because every CRD name in kubernetes follows this pattern <plural>.<group>.
 	group := regexp.MustCompile(groupMatch).FindStringSubmatch(crd.Name)[1]
-	if validation.CheckCRDGroup(group) && !validation.ValidateUserForCRD(v.whiteListedUsers, req.UserInfo) {
-		return admission.Denied(fmt.Sprintf("failed to validate user: %s in groups: %v to modify fleet CRD: %s", req.UserInfo.Username, req.UserInfo.Groups, crd.Name))
+	if validation.CheckCRDGroup(group) && !validation.IsMasterGroupUserOrWhiteListedUser(v.whiteListedUsers, req.UserInfo) {
```
Comment (ryanzhang-oss) [inline pkg/webhook/fleetresourcehandler/fleetresourcehandler_webhook.go:81]: why not make CRD the same as others and hide the internal details all in one validation call?

Comment (ryanzhang-oss) [reply pkg/webhook/fleetresourcehandler/fleetresourcehandler_webhook.go:81]: why delete the ValidateUserForFleetCRD?

File: pkg/webhook/validation/uservalidation.go [hunk 3]
```diff
@@ -2,27 +2,27 @@ package validation
...
 )
 
-// ValidateUserForCRD checks to see if user is authenticated to make a request to modify fleet CRDs.
-func ValidateUserForCRD(whiteListedUsers []string, userInfo authenticationv1.UserInfo) bool {
-	return isMasterGroupUserOrWhiteListedUser(whiteListedUsers, userInfo)
-}
-
-// ValidateUserForFleetCR checks to see if user is authenticated to make a request to modify Fleet CRs.
+// ValidateUserForFleetCR checks to see if user is allowed to make a request to modify fleet CRs.
 func ValidateUserForFleetCR(ctx context.Context, client client.Client, whiteListedUsers []string, userInfo authenticationv1.UserInfo) bool {
```
[the remark this answers is not in the document]
Comment (ryanzhang-oss) [reply pkg/webhook/validation/uservalidation.go:24]: actually member agent can't touch member Cluster CR at all

File: pkg/webhook/validation/uservalidation.go [hunk 4]
```diff
@@ -38,6 +38,21 @@ func ValidateUserForFleetCR(ctx context.Context, client client.Client, whiteList
...
+	}
+	return admission.Denied(fmt.Sprintf("user: %s in groups: %v is not allowed to modify fleet resource %s: %s/%s", userInfo.Username, userInfo.Groups, resKind, resName, resNamespace))
+}
+
+// IsMasterGroupUserOrWhiteListedUser returns true is user belongs to white listed users or user belongs to system:masters group.
+func IsMasterGroupUserOrWhiteListedUser(whiteListedUsers []string, userInfo authenticationv1.UserInfo) bool {
 	return slices.Contains(whiteListedUsers, userInfo.Username) || slices.Contains(userInfo.Groups, mastersGroup)
 }
+
+// IsUserAuthenticatedServiceAccount returns true if user is a valid/authenticated service account.
```
Comment (ryanzhang-oss) [inline pkg/webhook/validation/uservalidation.go:55]: I have a hunch that this should always be allowed for any resources

File: pkg/webhook/validation/uservalidation.go [hunk 5]
```diff
@@ -38,6 +38,21 @@ func ValidateUserForFleetCR(ctx context.Context, client client.Client, whiteList
...
+func IsUserAuthenticatedServiceAccount(userInfo authenticationv1.UserInfo) bool {
+	return slices.Contains(userInfo.Groups, serviceAccountsKubeSystemGroup) && slices.Contains(userInfo.Groups, serviceAccountsGroup) && slices.Contains(userInfo.Groups, authenticatedGroup)
```
Comment (michaelawyu) [inline pkg/webhook/validation/uservalidation.go:57]: Hi Arvind! This is an AND connected condition, but

any service account under the kube-system namespace will automatically be in the system:serviceaccounts group and system:authenticated group, so checking for system:serviceaccounts:kube-system alone should suffice

Comment (michaelawyu) [reply pkg/webhook/validation/uservalidation.go:57]: And any service account will automatically be in the system:authenticated group, so checking for system:serviceaccounts group should suffice

Comment (michaelawyu) [reply pkg/webhook/validation/uservalidation.go:57]: Is the purpose of this function to allow only service accounts in the kube-system namespace?

File: pkg/webhook/validation/uservalidation.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Arvindthiru) [reply pkg/webhook/validation/uservalidation.go]: Will keep in mind when implementing the webhook for IMC

File: pkg/webhook/validation/uservalidation.go [hunk 6]
```diff
@@ -1,43 +1,71 @@
...
+	fleetCRDGroups = []string{"networking.fleet.azure.com", "fleet.azure.com", "multicluster.x-k8s.io", "placement.azure.com"}
+)
 
-// ValidateUserForFleetCR checks to see if user is authenticated to make a request to modify Fleet CRs.
-func ValidateUserForFleetCR(ctx context.Context, client client.Client, whiteListedUsers []string, userInfo authenticationv1.UserInfo) bool {
-	if isMasterGroupUserOrWhiteListedUser(whiteListedUsers, userInfo) {
-		return true
+// ValidateUserForFleetCRD checks to see if user is not allowed to modify fleet CRDs.
+func ValidateUserForFleetCRD(group string, namespacedName types.NamespacedName, whiteListedUsers []string, userInfo authenticationv1.UserInfo) admission.Response {
+	if !checkCRDGroup(group) && isMasterGroupUserOrWhiteListedUser(whiteListedUsers, userInfo) {
```
Comment (michaelawyu) [inline pkg/webhook/validation/uservalidation.go:29]: Hi Arvind! Should this be an OR term?

**Review by michaelawyu — COMMENTED** [review]
Other than the comment things LGTM.

File: pkg/webhook/validation/uservalidation.go (hunk 6 above)
Comment (Arvindthiru) [reply pkg/webhook/validation/uservalidation.go]: I've updated the condition to only allow master group users and whitelisted users to modfiy fleet CRDs. We allow any user to modify other CRDs created by the user

File: pkg/webhook/fleetresourcehandler/fleetresourcehandler_webhook.go [hunk 7]
```diff
@@ -41,51 +45,71 @@ type fleetResourceValidator struct {
...
 		return admission.Errored(http.StatusBadRequest, err)
 	}
-
 	// This regex works because every CRD name in kubernetes follows this pattern <plural>.<group>.
 	group := regexp.MustCompile(groupMatch).FindStringSubmatch(crd.Name)[1]
-	if validation.CheckCRDGroup(group) && !validation.ValidateUserForCRD(v.whiteListedUsers, req.UserInfo) {
-		return admission.Denied(fmt.Sprintf("failed to validate user: %s in groups: %v to modify fleet CRD: %s", req.UserInfo.Username, req.UserInfo.Groups, crd.Name))
-	}
-	return admission.Allowed(fmt.Sprintf("user: %s in groups: %v is allowed to modify CRD: %s", req.UserInfo.Username, req.UserInfo.Groups, crd.Name))
+	return validation.ValidateUserForFleetCRD(group, types.NamespacedName{Name: crd.Name, Namespace: crd.Namespace}, v.whiteListedUsers, req.UserInfo)
```
Comment (ryanzhang-oss) [inline pkg/webhook/fleetresourcehandler/fleetresourcehandler_webhook.go:82]: CRD has no namespace

File: pkg/webhook/validation/uservalidation.go [hunk 8]
```diff
@@ -1,43 +1,71 @@
...
 	}
-	identities := make([]string, len(memberClusterList.Items))
-	for i := range memberClusterList.Items {
-		identities = append(identities, memberClusterList.Items[i].Spec.Identity.Name)
+	return response
+}
+
+// ValidateUserForResource checks to see if user is allowed to modify argued fleet resource.
+func ValidateUserForResource(resKind string, namespacedName types.NamespacedName, whiteListedUsers []string, userInfo authenticationv1.UserInfo) admission.Response {
+	if isMasterGroupUserOrWhiteListedUser(whiteListedUsers, userInfo) || isUserAuthenticatedServiceAccount(userInfo) {
```
Comment (ryanzhang-oss) [inline pkg/webhook/validation/uservalidation.go:50]: looks like resKind is not used in the validation?

**Review by ryanzhang-oss — APPROVED** [review]

**Pull request merged**